util/topsql/stmtstats/stmtstats.go

```diff
@@ -55,19 +72,57 @@ func CreateStatementStats() *StatementStats {
 	return stats
 }
 
-// OnExecutionBegin implements StatementObserver.OnExecutionBegin.
-func (s *StatementStats) OnExecutionBegin(sqlDigest, planDigest []byte) {
+// OnParseBegin implements StatementObserver.OnParseBegin.
+func (s *StatementStats) OnParseBegin() {
+	s.seCtx.parseBeginUnixNano = nowFunc().UnixNano()
+}
+
+// OnSQLDigestReady implements StatementObserver.OnSQLDigestReady.
+func (s *StatementStats) OnSQLDigestReady(sqlDigest []byte) {
+	s.seCtx.sqlDigest.Store(sqlDigest)
+}
+
+// OnPlanDigestReady implements StatementObserver.OnParseBegin.
+func (s *StatementStats) OnPlanDigestReady(planDigest []byte) {
+	s.seCtx.planDigest.Store(planDigest)
+}
+
+// OnExecBegin implements StatementObserver.OnExecBegin.
+func (s *StatementStats) OnExecBegin() {
+	s.seCtx.execBeginUnixNano = nowFunc().UnixNano()
+}
+
+// OnExecFinished implements StatementObserver.OnExecFinished.
+func (s *StatementStats) OnExecFinished() {
+	sqlDigest, ok1 := s.seCtx.sqlDigest.Load().([]byte)
+	planDigest, ok2 := s.seCtx.planDigest.Load().([]byte)
```

Comment:
If `planDigest` is nil, will the `.([]byte)` panic here?

Reply:
`seCtx.planDigest` type is `atomic.Value`, it won't be nil. if the value of `seCtx.planDigest` stored is nil, it won't be panic here, but the `ok2` will be false.  Also there are some test call `OnPlanDigestReady(nil)` to store nil in `stmtstats_test.go`